### PR TITLE
Return field.Errors from node affinity parsing

### PR DESCRIPTION
**Pull request by alculquicondor**
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

Provides all and more detailed parsing errors.

**Which issue(s) this PR fixes**:

Part of #96167

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by alculquicondor** [comment]
/triage accepted
/assign @damemi 

cc @lingsamuel

**Comment by alculquicondor** [comment]
/assign @liggitt 
for mod file and sanity check on API validation.

File: staging/src/k8s.io/component-helpers/scheduling/corev1/nodeaffinity/nodeaffinity.go [hunk 1]
```diff
@@ -37,31 +38,44 @@ type LazyErrorNodeSelector struct {
...
 // errors when no terms match.
-func NewLazyErrorNodeSelector(ns *v1.NodeSelector) *LazyErrorNodeSelector {
+func NewLazyErrorNodeSelector(ns *v1.NodeSelector, opts ...ParseOption) *LazyErrorNodeSelector {
+	o := parseOptions{}
+	for _, opt := range opts {
+		opt(&o)
+	}
 	parsedTerms := make([]nodeSelectorTerm, 0, len(ns.NodeSelectorTerms))
-	for _, term := range ns.NodeSelectorTerms {
+	path := o.path.Child("nodeSelectorTerms")
```
Comment (damemi) [inline staging/src/k8s.io/component-helpers/scheduling/corev1/nodeaffinity/nodeaffinity.go:71]: `o.path` could be nil right?

File: staging/src/k8s.io/component-helpers/scheduling/corev1/nodeaffinity/nodeaffinity.go [hunk 2]
```diff
@@ -107,25 +121,30 @@ type PreferredSchedulingTerms struct {
...
+func NewPreferredSchedulingTerms(terms []v1.PreferredSchedulingTerm, opts ...ParseOption) (*PreferredSchedulingTerms, error) {
+	o := parseOptions{}
+	for _, opt := range opts {
+		opt(&o)
+	}
+	var errs field.ErrorList
 	parsedTerms := make([]preferredSchedulingTerm, 0, len(terms))
-	for _, term := range terms {
+	for i, term := range terms {
+		path := o.path.Index(i)
```
Comment (damemi) [inline staging/src/k8s.io/component-helpers/scheduling/corev1/nodeaffinity/nodeaffinity.go:132]: same question

**Review by damemi — COMMENTED** [review]
looks pretty good but I think you need to check if `path` is nil in the spots you're using it (but correct me if I'm wrong 🙂)

File: staging/src/k8s.io/component-helpers/scheduling/corev1/nodeaffinity/nodeaffinity.go (hunk 1 above)
Comment (alculquicondor) [reply staging/src/k8s.io/component-helpers/scheduling/corev1/nodeaffinity/nodeaffinity.go:71]: Good eye, but `(nil)`.Child actually works :)

File: pkg/scheduler/apis/config/validation/validation_pluginargs.go [hunk 3]
```diff
@@ -265,22 +266,35 @@ func ValidateNodeAffinityArgs(args *config.NodeAffinityArgs) error {
...
 		}
 	}
 	// TODO: Add validation for requiredDuringSchedulingRequiredDuringExecution when it gets added to the API.
 	if terms := affinity.PreferredDuringSchedulingIgnoredDuringExecution; len(terms) != 0 {
-		_, err := nodeaffinity.NewPreferredSchedulingTerms(terms)
+		_, err := nodeaffinity.NewPreferredSchedulingTerms(terms, nodeaffinity.WithPath(path.Child("preferredDuringSchedulingIgnoredDuringExecution")))
 		if err != nil {
-			// TODO(#96167): Expand all field.Error(s) returned once constructor use them.
-			allErrs = append(allErrs, field.Invalid(f.Child("preferredDuringSchedulingIgnoredDuringExecution"), affinity.PreferredDuringSchedulingIgnoredDuringExecution, err.Error()))
+			allErrs = appendFieldErrorListFromErr(allErrs, err)
```
Comment (liggitt) [inline pkg/scheduler/apis/config/validation/validation_pluginargs.go:281]: could probably just keep appending err as you were, and then flatten the aggregate error at the end, without needing `appendFieldErrorListFromErr` or panics

```
return errors.Flatten(allErrs.ToAggregate())
```

File: pkg/scheduler/apis/config/validation/validation_pluginargs.go [hunk 4]
```diff
@@ -265,22 +266,35 @@ func ValidateNodeAffinityArgs(args *config.NodeAffinityArgs) error {
...
+	agg, ok := err.(errors.Aggregate)
+	if !ok {
+		panic("err is not an errors.Aggregate")
+	}
+	for _, e := range agg.Errors() {
+		fErr, ok := e.(*field.Error)
+		if !ok {
+			panic("wrapped err is not a field.Err")
+		}
+		errs = append(errs, fErr)
```
Comment (liggitt) [inline pkg/scheduler/apis/config/validation/validation_pluginargs.go:297]: seems like we'd still be ok just appending errors we don't recognize as special types... is just flattening the aggregated error using `errors.Flatten` problematic?

Comment (alculquicondor) [reply pkg/scheduler/apis/config/validation/validation_pluginargs.go:297]: You are right, we shouldn't care about the type. I didn't know about `errors.Flatten`. It's very handy.

File: pkg/scheduler/apis/config/validation/validation_pluginargs.go (hunk 3 above)
Comment (alculquicondor) [reply pkg/scheduler/apis/config/validation/validation_pluginargs.go:281]: Done

**Review by damemi — APPROVED** [review]
/lgtm

**Comment by liggitt** [comment]
/approve
for dep and validation structural change

/hold 
this is past code freeze, and it doesn't look like this is an urgent release blocking bug. I'd suggest this move to 1.21.

**Comment by dims** [comment]
Ack. moving to v1.21

/milestone v1.21

**Comment by alculquicondor** [comment]
/hold

it looks like there are some old dependencies. I will unhold once I rebase.

**Comment by alculquicondor** [comment]
/hold cancel

@damemi PTAL
just updated cmp vendor

**Pull request merged**